### PR TITLE
Added backward function for kl_div target

### LINKED ISSUE #5081: [feature request] Enable backprop through 'target' in F.kl_div
Linked from: https://discuss.pytorch.org/t/runtimeerror-the-derivative-for-target-is-not-implemented-when-trying-to-backprop-through-kl/13362

As in the title, it would be really useful to be able to backprop through F.kl_div in both arguments. Calculating the KL divergence between two distributions is fairly common thing to do during the forward pass of a model (i.e. as a form of regularisation), and having a C-optimized way of doing so would be very convenient, if you could just differentiate through the second argument!

Thanks
Kris

### PR DESCRIPTION
**Pull request by weiyangfb**
Summary:
1. Added backward function for kl_div target ATen
2. Added grad test for target in test_nn.py

This fixes #5081

### CODE REVIEW

File: test/test_nn.py [hunk 1]
```diff
@@ -377,6 +377,7 @@ class NewCriterionTest(InputVariableMixin, CriterionTest):
     def __init__(self, *args, **kwargs):
         super(NewCriterionTest, self).__init__(*args, **kwargs)
         self.check_gradgrad = kwargs.get('check_gradgrad', True)
+        self.check_grad_target = kwargs.get('check_grad_target', False)
```
Comment (zou3519) [inline test/test_nn.py]: I'm not sure we want to add more functionality to the CriterionTest because we're moving away from a notion of a "loss function" -- all losses should be modules. 

You should be able to write a [module test](https://github.com/pytorch/pytorch/blob/master/test/test_nn.py#L6025-L6035) for this. All module tests test grad of their inputs (and gradgrad), so you should be able to send in a target with requires_grad=True and see if it works.

File: tools/autograd/templates/Functions.cpp [hunk 2]
```diff
@@ -665,6 +665,18 @@ Tensor kl_div_double_backward_grad_output(const Tensor & grad, const Tensor & in
...
+Tensor kl_div_target_backward(Tensor grad_output, Tensor self, Tensor target, bool size_average, bool reduce) {
+  if (!reduce) {
+    return grad_output.mul(target.log() + at::ones_like(target) - self).masked_fill_(target == 0, 0.);
+  }
+
+  auto norm = at::ones_like(target);
+  if (size_average) {
+    norm = norm.div(norm.numel());
+  }
+  return grad_output.mul(target.log() + at::ones_like(target) - self).mul(norm).masked_fill_(target == 0, 0.);
```
Comment (zou3519) [inline tools/autograd/templates/Functions.cpp]: Is it possible to avoid the allocation of a `norm` tensor? You might be able to allocate a scalar tensor for it, or just use a scalar entirely: (does this work?)
```
(grad_output.mul(target.log() + at::ones_like(target) - self) / grad_output.numel()).masked_fill(...)
```

File: tools/autograd/derivatives.yaml [hunk 3]
```diff
@@ -963,6 +964,7 @@
 - name: kl_div_backward(Tensor grad_output, Tensor self, Tensor target, bool size_average, bool reduce)
   grad_output: kl_div_double_backward_grad_output(grad, self, target, size_average, reduce)
   self: zeros_like(grad)
+  target: zeros_like(grad)
```
Comment (zou3519) [inline tools/autograd/derivatives.yaml]: What does this do?

File: tools/autograd/templates/Functions.cpp [hunk 4]
```diff
@@ -665,6 +665,18 @@ Tensor kl_div_double_backward_grad_output(const Tensor & grad, const Tensor & in
   return result;
 }
 
+Tensor kl_div_target_backward(Tensor grad_output, Tensor self, Tensor target, bool size_average, bool reduce) {
+  if (!reduce) {
+    return grad_output.mul(target.log() + at::ones_like(target) - self).masked_fill_(target == 0, 0.);
```
Comment (SsnL) [inline tools/autograd/templates/Functions.cpp]: how about just `.add_(1).sub_(self)`? saving one allocation.

File: tools/autograd/derivatives.yaml (hunk 3 above)
Comment (weiyangfb) [reply tools/autograd/derivatives.yaml]: oops, will remove it. Yes, it was for double-backward. But sounds like I don't need that.

Comment (SsnL) [reply tools/autograd/derivatives.yaml]: If it computes the correct ggTarget, why should we remove it?

Comment (weiyangfb) [reply tools/autograd/derivatives.yaml]: ah, I see, it makes sense, I will keep it then.

File: tools/autograd/derivatives.yaml [hunk 5]
```diff
@@ -718,6 +718,7 @@
 
 - name: kl_div_forward(Tensor self, Tensor target, bool size_average, bool reduce)
   self: kl_div_backward(grad, self, target, size_average, reduce)
+  target: kl_div_target_backward(grad, self, target, size_average, reduce)
```
[the remark this answers is not in the document]
Comment (weiyangfb) [reply tools/autograd/derivatives.yaml]: No, it is not trying to avoid ggTarget. It is simply that we can't find actual implementation of kl_div_backward, but only its generated files.

[the remark this answers is not in the document]
Comment (zou3519) [reply tools/autograd/derivatives.yaml]: Adding on, kl_div_backward is generated from TH

[the remark this answers is not in the document]
Comment (weiyangfb) [reply tools/autograd/derivatives.yaml]: I still can't find it, I must have missed something. So I can see the generated files: ./aten/build/src/ATen/ATen/CUDADoubleType.cpp, and it uses fn defined in ./aten/src/THNN/generic/DistKLDivCriterion.c

[the remark this answers is not in the document]
Comment (SsnL) [reply tools/autograd/derivatives.yaml]: It is generated from `nn.yaml`. If we want to have double backward for gTarget in this PR, one way we can do is to make the new method wrap around `kl_div_backward`.

File: test/test_nn.py (hunk 1 above)
Comment (weiyangfb) [reply test/test_nn.py]: I see, I will add a new test for kl_div target as `module test`. Is there a way to pytest all/part of NewModuleTest only?

Comment (zou3519) [reply test/test_nn.py]: Like test only specific tests? You can do `pytest test/test_nn.py -k "kl_div" -v` (if you have pytest installed) to test all tests in test_nn.py that have the string "kl_div" in them.

Comment (weiyangfb) [reply test/test_nn.py]: Right, that's exactly what I wanted! But looks like pytest only runs on methods with prefix `test_`. Currently I have the following results if I do `pytest test/test_nn.py -k "kl_div" -v` or `pytest test/test_nn.py -k "kldiv" -v`

```
pytest test/test_nn.py -k "kl_div" -v
================================================================= test session starts ==================================================================
platform linux2 -- Python 2.7.14, pytest-3.3.2, py-1.5.2, pluggy-0.6.0 -- /home/weiyang/anaconda2/bin/python
cachedir: .cache
rootdir: /home/weiyang/pytorch, inifile:
collected 1128 items

=================================================================== warnings summary ===================================================================
test/test_nn.py
  cannot collect 'test' because it is not a function.

test/test_nn.py::TestBase
  cannot collect test class 'TestBase' because it has a __init__ constructor

-- Docs: http://doc.pytest.org/en/latest/warnings.html
```

Comment (zou3519) [reply test/test_nn.py]: Yikes, pytest is case sensitive, try: `pytest test/test_nn.py -k "KLDiv"  -v`

Comment (weiyangfb) [reply test/test_nn.py]: Bingo! It works like a charm! Thanks a ton!

File: tools/autograd/derivatives.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (ezyang) [reply tools/autograd/derivatives.yaml]: I don't mind if we don't actually manage to call THNN. We wanna kill it all anyway. Might be worth adding a note to the impl in THNN though, saying that is dead.

**Review by ezyang — APPROVED** [review]
